### DISCUSSION TITLE
How can I send error details to the client?

### DISCUSSION THREAD

**Discussion by kamyar1979** [icerpc/icerpc-csharp#3948] [Q&A] [answered]
We are using Redis for invocation: In the client side, we have implemented IInvoker and put the OutgoingRequest into Redis stream, and the server reads the IncomingRequest from Redis stream and uses DispatchAsync to call the corresponding method. The problem is when we raise an exception within the method, DispatchAsync raises the exception on the server side, but we expect it to create an OutgoingResponse containing the error StatusCode and ErrorMessage so that we can put in Redis, and the error raises on the client side. Is there a way to do it?

**Comment by bernardnormier** [comment, marked as answer]
Hi Kamyar,

So you've essentially implemented a new "redis" protocol, alongside the `ice` and `icerpc` protocols that IceRPC provides. I didn't anticipate replacing the protocol would be a common use-case.

> Is there a way to do it?

Yes, you can easily create an OutgoingResponse from a DispatchException. That's what IceRpcProtocolConnection does here:
https://github.com/icerpc/icerpc-csharp/blob/c62195befeba8c9729d3700046ca8d78b1fb13ad/src/IceRpc/Internal/IceRpcProtocolConnection.cs#L1221

`ToOutgoingResponse` is an internal extension method:
https://github.com/icerpc/icerpc-csharp/blob/c62195befeba8c9729d3700046ca8d78b1fb13ad/src/IceRpc/Internal/DispatchExceptionExtensions.cs#L6

Maybe you could open a PR to convert it into a public extension method?

All the best,
Bernard

**Reply by kamyar1979** [reply]
Not really! I guess we have implemented a Redis transport since we are using IceRPC protocol. 

`                using var request =
                    new IncomingRequest(IceRpc.Protocol.IceRpc, FakeConnectionContext.Instance)
                    {
                        IsOneway = false,
                        Operation = operation
                    };
                request.Payload =
                    PipeReader.Create(new ReadOnlySequence<byte>(cmd[PubSubKeys.Payload]!));
`

I checked the code and seems it is implemented within IceRPC protocol, which we are just using! I guess there is a way to make the ICeRPC protocol do it: I can wrap my own application exception within a DispatchException so that ToOutgoingResponse invokes.

**Comment by kamyar1979** [comment]
Wrapping the exception within DispatchException did the trick! Thank you!
